File: selected_coding_interview/codes/cpp/lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list/lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list_s2.cpp
/*
 * File: lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list_s2.cpp
 * Created Time: 2025-12-30
 * Author: krahets
 */

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    Node* treeToDoublyList(Node* root) {
        if (root == nullptr) return nullptr;
        dfs(root);
        head->left = pre;
        pre->right = head;
        return head;
    }

private:
    Node *pre, *head;
    void dfs(Node* cur) {
        if (cur == nullptr) return;
        dfs(cur->left);
        if (pre != nullptr) pre->right = cur;
        else head = cur;
        cur->left = pre;
        pre = cur;
        dfs(cur->right);
    }
};

int main() {
    // ======= Test Case =======
    // TODO: Add specific test case

    // ====== Driver Code ======
    Solution* slt = new Solution();
    // TODO: Call solution method and print result

    return 0;
}

File: selected_coding_interview/codes/java/lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list/lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list_s2.java
/*
* File: lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list_s2.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        Node pre, head;
        void dfs(Node cur) {
            if (cur == null) return;
            dfs(cur.left);
            if (pre != null) pre.right = cur;
            else head = cur;
            cur.left = pre;
            pre = cur;
            dfs(cur.right);
        }
        public Node treeToDoublyList(Node root) {
            if (root == null) return null;
            dfs(root);
            head.left = pre;
            pre.right = head;
            return head;
        }
    }

public class lc_426_convert_binary_search_tree_to_sorted_doubly_linked_list_s2 {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        // TODO: Add appropriate test case

        // ====== Driver Code ======
        Solution slt = new Solution();
        // result = slt.treeToDoublyList(...)
        // print(result)

    }
}
